File: day07/day07.zig
const std = @import("std");

const MemoKey = struct {
    row: usize,
    col: usize,
};

pub fn main() !void {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    defer _ = gpa.deinit();
    const allocator = gpa.allocator();

    const file = try std.fs.cwd().openFile("input.txt", .{});
    defer file.close();

    const content = try file.readToEndAlloc(allocator, 10 * 1024 * 1024);
    defer allocator.free(content);

    // Parse lines
    var line_list = std.ArrayListUnmanaged([]const u8){};
    defer line_list.deinit(allocator);

    var lines_iter = std.mem.splitScalar(u8, content, '\n');
    while (lines_iter.next()) |line| {
        if (line.len > 0) {
            try line_list.append(allocator, line);
        }
    }

    const lines = line_list.items;
    const height = lines.len;

    // Find start column
    var start_col: usize = 0;
    for (lines[0], 0..) |c, i| {
        if (c == 'S') {
            start_col = i;
            break;
        }
    }

    // Part 1: Count beam splits (beams merge at same position)
    const part1_result = blk: {
        var beams = std.AutoHashMap(usize, void).init(allocator);
        defer beams.deinit();
        try beams.put(start_col, {});

        var splits: u32 = 0;

        for (0..height - 1) |row| {
            const next_row = lines[row + 1];

            var new_beams = std.AutoHashMap(usize, void).init(allocator);

            var iter = beams.keyIterator();
            while (iter.next()) |col_ptr| {
                const col = col_ptr.*;
                if (next_row[col] == '^') {
                    splits += 1;
                    try new_beams.put(col - 1, {});
                    try new_beams.put(col + 1, {});
                } else {
                    try new_beams.put(col, {});
                }
            }

            beams.deinit();
            beams = new_beams;
        }

        break :blk splits;
    };

    // Part 2: Count timelines (many-worlds interpretation with memoization)
    var memo = std.AutoHashMap(MemoKey, i64).init(allocator);
    defer memo.deinit();

    const part2_result = try timelines(0, start_col, lines, height, &memo);

    std.debug.print("Part 1: {d}\n", .{part1_result});
    std.debug.print("Part 2: {d}\n", .{part2_result});
}

fn timelines(
    row: usize,
    col: usize,
    lines: []const []const u8,
    height: usize,
    memo: *std.AutoHashMap(MemoKey, i64),
) !i64 {
    if (row == height - 1) {
        return 1;
    }

    const key = MemoKey{ .row = row, .col = col };
    if (memo.get(key)) |val| {
        return val;
    }

    const result: i64 = if (lines[row + 1][col] == '^')
        try timelines(row + 1, col - 1, lines, height, memo) +
            try timelines(row + 1, col + 1, lines, height, memo)
    else
        try timelines(row + 1, col, lines, height, memo);

    try memo.put(key, result);
    return result;
}

File: day07/day07.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strings"
)

type Pos struct {
	row, col int
}

// Part 1: Count beam splits (beams merge at same position)
func part1(lines []string) int {
	height := len(lines)
	startCol := strings.IndexByte(lines[0], 'S')

	beams := map[int]bool{startCol: true}
	splits := 0

	for row := 0; row < height-1; row++ {
		nextRow := lines[row+1]
		newBeams := make(map[int]bool)

		for col := range beams {
			if nextRow[col] == '^' {
				splits++
				newBeams[col-1] = true
				newBeams[col+1] = true
			} else {
				newBeams[col] = true
			}
		}
		beams = newBeams
	}

	return splits
}

// Part 2: Count timelines (many-worlds interpretation with memoization)
func part2(lines []string) int64 {
	height := len(lines)
	startCol := strings.IndexByte(lines[0], 'S')

	memo := make(map[Pos]int64)

	var timelines func(row, col int) int64
	timelines = func(row, col int) int64 {
		if row == height-1 {
			return 1
		}

		pos := Pos{row, col}
		if val, ok := memo[pos]; ok {
			return val
		}

		var result int64
		if lines[row+1][col] == '^' {
			result = timelines(row+1, col-1) + timelines(row+1, col+1)
		} else {
			result = timelines(row+1, col)
		}

		memo[pos] = result
		return result
	}

	return timelines(0, startCol)
}

func main() {
	file, err := os.Open("input.txt")
	if err != nil {
		fmt.Fprintf(os.Stderr, "Failed to open input.txt: %v\n", err)
		os.Exit(1)
	}
	defer file.Close()

	var lines []string
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := scanner.Text()
		if len(line) > 0 {
			lines = append(lines, line)
		}
	}

	if err := scanner.Err(); err != nil {
		fmt.Fprintf(os.Stderr, "Error reading file: %v\n", err)
		os.Exit(1)
	}

	fmt.Printf("Part 1: %d\n", part1(lines))
	fmt.Printf("Part 2: %d\n", part2(lines))
}
